pydantic/main.py

```diff
@@ -794,7 +790,7 @@ def __class_getitem__(
                 # Attempt to rebuild the origin in case new types have been defined
                 try:
                     # depth 3 gets you above this __class_getitem__ call
-                    origin.model_rebuild(_parent_namespace_depth=3)
+                    origin.model_rebuild(_parent_namespace_depth=0)
```

Comment:
Why the depth change here? We should change the comment...

Reply:
Ah yes forgot to explain this. When you do the following:

```python
class Model(BaseModel, Generic[T]):
    a: 'Forward'

Forward = int

ModelInt = Model[int]
```

Doing `Model[int]` will trigger this `origin.model_rebuild` call (because `'Forward'` failed to resolve initially). While it may make sense to fetch the parent namespace here, it doesn't in this case:

```python
class OtherModel(BaseModel):
    m: 'Model[int]'
```

Because when the annotation for `m` is going to be evaluated, we will call `eval('Model[int]', some_globals, some_lazy_local_ns_instance)`. A new frame will be created, with `f_locals = some_lazy_local_ns_instance`. `Model[int]` will trigger the `origin.model_rebuild` call, and fetching the parent namespace will resolve to `some_lazy_local_ns_instance`: highly unexpected!

This caused a bug with one test case:

```python
from typing import Generic, TypeVar, Union

from pydantic import BaseModel

T1 = TypeVar('T1')
T2 = TypeVar('T2')
T3 = TypeVar('T3')

class A1(BaseModel, Generic[T1]):
    a1: 'A2[T1]'

class A2(BaseModel, Generic[T2]):
    a2: 'A3[T2]'

class A3(BaseModel, Generic[T3]):
    a3: Union['A1[T3]', T3]
```

Because we now pass the `ns_resolver` to `get_cls_type_hints` in `collect_model_fields` (TODO, should we? Let me investigate), and because this `ns_resolver` will add `{A3.__name__: A3}` to the locals for every type: when collecting fields for `A3`, we rebuild `A1` (because `'A1[T3]'` is evaluated), which rebuilds `A2`, which rebuilds `A3` (but fields for `A3` were not even collected yet! Resulting in an attribute error because it tries to access `A3.__pydantic_fields__` somewhere).

This needs further investigation.